ddl/index.go

```diff
@@ -161,35 +162,39 @@ func checkIndexColumn(col *model.ColumnInfo, ic *ast.IndexColName) error {
 }
 
 func getIndexColumnLength(col *model.ColumnInfo, colLen int) (int, error) {
-	// Take care of the sum of length of all index columns.
+	length := types.UnspecifiedLength
 	if colLen != types.UnspecifiedLength {
-		return colLen, nil
-	}
-	// Specified data types.
-	if col.Flen != types.UnspecifiedLength {
-		// Special case for the bit type.
-		if col.FieldType.Tp == mysql.TypeBit {
-			return (col.Flen + 7) >> 3, nil
-		}
-		return col.Flen, nil
-
+		length = colLen
+	} else if col.Flen != types.UnspecifiedLength {
+		length = col.Flen
 	}
 
-	length, ok := mysql.DefaultLengthOfMysqlTypes[col.FieldType.Tp]
-	if !ok {
-		return length, errUnknownTypeLength.GenWithStackByArgs(col.FieldType.Tp)
-	}
-
-	// Special case for time fraction.
-	if types.IsTypeFractionable(col.FieldType.Tp) &&
-		col.FieldType.Decimal != types.UnspecifiedLength {
-		decimalLength, ok := mysql.DefaultLengthOfTimeFraction[col.FieldType.Decimal]
-		if !ok {
-			return length, errUnknownFractionLength.GenWithStackByArgs(col.FieldType.Tp, col.FieldType.Decimal)
-		}
-		length += decimalLength
+	switch col.Tp {
+	case mysql.TypeBit:
+		return (length + 7) >> 3, nil
+	case mysql.TypeVarchar, mysql.TypeString:
+		desc, err := charset.GetCharsetDesc(col.Charset)
+		if err != nil {
+			return 0, errUnsupportedCharset.GenWithStackByArgs(col.Charset, col.Collate)
+		}
+		return desc.Maxlen * length, nil
+	case mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeBlob, mysql.TypeLongBlob:
+		return length, nil
+	case mysql.TypeTiny, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeDouble, mysql.TypeShort:
+		return mysql.DefaultLengthOfMysqlTypes[col.Tp], nil
+	case mysql.TypeFloat:
+		if length <= mysql.MaxFloatPrecisionLength {
+			return mysql.DefaultLengthOfMysqlTypes[mysql.TypeFloat], nil
+		}
+		return mysql.DefaultLengthOfMysqlTypes[mysql.TypeDouble], nil
+	case mysql.TypeDecimal, mysql.TypeNewDecimal:
+		// Decimal using a binary format that packs nine decimal (base 10) digits into four bytes.
+		return (length / 9 * 4) + ((length%9)+1)/2, nil
```

Comment:
Could you put it into a function? Thus we can understand getIndexColumnLength easily